webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that "a STOP_SENDING frame requests that the receiving endpoint send a
+RESET_STREAM frame.". See [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5).
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
```

Comment:
I find it confusing to have a section about ordering here. While possible in principle, we don’t use unordered abstractions anywhere in libp2p.

Reply:
I think it is fine to keep it in here, given that the option exists and implementers might ask themselves whether they SHOULD support it. In case you feel strongly about it @marten-seemann I am happy to remote it.

Reply:
Resolving here. Again, in case you feel strongly about removing this, please speak up.